### PR TITLE
SPICE-0028: Add support for multi-line string line continuations

### PR DESCRIPTION
**Pull request by HT154**
SPICE: https://github.com/apple/pkl-evolution/pull/31

### CODE REVIEW

File: docs/modules/language-reference/pages/index.adoc [hunk 1]
```diff
@@ -305,7 +304,8 @@ To write a string that spans multiple lines, use a multiline string literal:
 ----
 """
 Although the Dodo is extinct,
-the species will be remembered.
+the species \
+will be remembered.
```
Comment (bioball) [inline docs/modules/language-reference/pages/index.adoc]: This is the first time multiline strings are introduced. I think it's better to talk about continuations elsewhere.

I think we can have another paragraph at the bottom of this section, something like:

```asciidoc
To prevent line breaks from becoming part of the string's value, use a backslash (`\`) to end those lines.

[source%tested,{pkl}]
----
str = """
  Although the Dodo is extinct, \
  the species will be remembered.
  """
----

This multiline string is equivalent to the following single-line string:

[source%parsed,{pkl-expr}]
----
"Although the Dodo is extinct, the species will be remembered."
----
```

**Review by bioball — CHANGES_REQUESTED** [review]
Looks good! Just some minor comments.

Can we also add a parser error test for whitespace after line continuation?

File: pkl-core/src/test/files/LanguageSnippetTests/input/basic/stringMultiline.pkl [hunk 2]
```diff
@@ -25,6 +25,22 @@ examples {
...
+    #"""
+    hello \#
+    world
+    """#
+  
+    """
+    hello \
+    \
+    world
+    """
```
Comment (bioball) [inline pkl-core/src/test/files/LanguageSnippetTests/input/basic/stringMultiline.pkl]: Move these into a new entry called "string continuation"

File: pkl-core/src/test/files/LanguageSnippetTests/output/errors/parser21.err [hunk 3]
```diff
@@ -0,0 +1,8 @@
+–– Pkl Error ––
+Invalid character escape sequence `\ `.
+
+Valid character escape sequences are:  \n  \r  \t  \"  \\
```
Comment (bioball) [inline pkl-core/src/test/files/LanguageSnippetTests/output/errors/parser21.err]: Can we add some sort of hint that string continuation is also valid?

**Review by bioball — APPROVED** [review]
LGTM!

Also looking for a thumb from @stackoverflow

File: pkl-core/src/test/files/LanguageSnippetTests/output/errors/parser21.err (hunk 3 above)
Comment (HT154) [reply pkl-core/src/test/files/LanguageSnippetTests/output/errors/parser21.err]: Hmm given that `\n` is already present here, any ideas on how to communicate this? `\«newline»`?

Comment (HT154) [reply pkl-core/src/test/files/LanguageSnippetTests/output/errors/parser21.err]: Went with `\<newline>`, but open to input here

File: pkl-formatter/src/main/java/org/pkl/formatter/Builder.java [hunk 4]
```diff
@@ -922,7 +927,9 @@ private List<FormatNode> formatStringParts(List<Node> nodes) {
       if (elem.type == NodeType.TERMINAL && text(elem).endsWith("(")) {
         isInStringInterpolation = true;
       }
-      result.add(format(elem));
+      var formatted = format(elem);
+      if (formatted instanceof Nodes formattedNodes) result.addAll(formattedNodes.nodes());
+      else result.add(format(elem));
```
Comment (stackoverflow) [inline pkl-formatter/src/main/java/org/pkl/formatter/Builder.java]:
```suggestion
      else result.add(formatted);
```

**Review by stackoverflow — APPROVED** [review]
LGTM.

File: pkl-formatter/src/main/java/org/pkl/formatter/Builder.java (hunk 4 above)
Comment (HT154) [reply pkl-formatter/src/main/java/org/pkl/formatter/Builder.java]: Ah good catch, thanks! Must've missed this one with the kt->java translation.
